### ISSUE TITLE
systemz: "error in backend: Unsupported stack frame traversal count"

### ISSUE THREAD

**Issue by 9816046d-d175-4de2-95d1-ce389e5190e8** [llvm/llvm-project#40769]
|  |  |
| --- | --- |
| Bugzilla Link | [41424](https://llvm.org/bz41424) |
| Version | 8.0 |
| OS | Linux |
| Blocks | llvm/llvm-project#4440  |
| Attachments | [preprocessed kernelhttps://user-images.githubusercontent.com/92601431/143759005-1758b79f-c88b-4249-95d2-0271749bd8b3.gz), [reproducer sript for sched_wakeup.c](https://user-images.githubusercontent.com/92601431/143759006-45b0f126-a53f-4f6d-b56b-808683cdfa55.gz) |
| CC | @nickdesaulniers,@uweigand |

## Extended Description 

I get a build failure for various file when trying to compile the linux kernel for s390 (systemz):

clang-8" "-cc1" "-triple" "s390x-unknown-linux-gnu" "-S" "-disable-free" "-disable-llvm-verifier" "-discard-value-names" "-main-file-name" "trace_sched_wakeup.c" "-mrelocation-model" "static" "-mthread-model" "posix" "-relaxed-aliasing" "-fmath-errno" "-masm-verbose" "-no-integrated-as" "-mconstructor-aliases" "-fuse-init-array" "-target-cpu" "z196" "-mbackchain" "-dwarf-column-info" "-debug-info-kind=limited" "-dwarf-version=4" "-debugger-tuning=gdb" "-momit-leaf-frame-pointer" "-coverage-notes-file" "/git/arm-soc/y2038/obj-s390/kernel/trace/trace_sched_wakeup.gcno" "-nostdsysteminc" "-nobuiltininc" "-sys-header-deps" "-D" "__KERNEL__" "-D" "CONFIG_AS_CFI_VAL_OFFSET=1" "-D" "CC_USING_FENTRY" "-D" "KBUILD_BASENAME=\"trace_sched_wakeup\"" "-D" "KBUILD_MODNAME=\"trace_sched_wakeup\"" "-O2" "-Wall" "-Wundef" "-Werror=strict-prototypes" "-Wno-trigraphs" "-Werror=implicit-function-declaration" "-Werror=implicit-int" "-Wno-format-security" "-Wno-sign-compare" "-Wno-format-invalid-specifier" "-Wno-gnu" "-Wno-address-of-packed-member" "-Wno-tautological-compare" "-Wno-unused-const-variable" "-Wdeclaration-after-statement" "-Wvla" "-Wno-pointer-sign" "-Werror=date-time" "-Werror=incompatible-pointer-types" "-Wno-initializer-overrides" "-Wno-unused-value" "-Wno-format" "-Wno-sign-compare" "-Wno-format-zero-length" "-std=gnu89" "-fno-dwarf-directory-asm" "-ferror-limit" "19" "-fmessage-length" "132" "-fwrapv" "-fno-signed-char" "-fwchar-type=short" "-fno-signed-wchar" "-fobjc-runtime=gcc" "-fno-common" "-fdiagnostics-show-option" "-fcolor-diagnostics" "-vectorize-loops" "-vectorize-slp" "-x" "c" "trace_sched_wakeup-ff17db.c"

fatal error: error in backend: Unsupported stack frame traversal count

I did not try reducing the test case, but could do so if it's not immediately clear from the output where the problem is.

**Comment by dominik-steenken** [comment]
I'm not sure what the exact intended resolution here is, but i think the support for `builtin_{frame,return}_address()` with non-zero argument with `-mbackchain` was added with https://github.com/llvm/llvm-project/pull/69405 and subsequent PRs https://github.com/llvm/llvm-project/pull/71668 and https://github.com/llvm/llvm-project/pull/71834.
It also seems to me that the kernel is now generally buildable with LLVM https://docs.kernel.org/kbuild/llvm.html.
@uweigand FYI

**Comment by uweigand** [comment]
Yes, I think this works well enough for now.  If there's any further requirements by the kernel (e.g. `-mkernel-backchain` support), this can be handled via new issues.
